### PR TITLE
Zero copy for Eigen::Matrix3XT and Eigen::Map<const Eigen::Matrix3XT>

### PR DESCRIPTION
**Pull request by simonlynen**
* Make nabo a template of the cloud type.
* Add explicit instantiation for some types:
  * ```Eigen::MatrixXf```
  * ```Eigen::MatrixXd```
  * ```Eigen::Matrix3Xf```
  * ```Eigen::Matrix3Xd```
  * ```Eigen::Map<const Matrix3Xf, Eigen::Aligned>```
  * ```Eigen::Map<const Eigen::Matrix3Xd, Eigen::Aligned>```

* Added helpers to test to allow testing these types.

* Fixed opencl implementation, but one method is not implemented and runtime crashes occur on my machine.

### CODE REVIEW

**Comment by simonlynen** [comment]
@HannesSommer @tcies ptal

**Comment by HannesSommer** [comment]
Oh my, this is pretty big... I couldn't do a thorough review. But for my time budget: lgtm.

**Comment by simonlynen** [comment]
@dymczykm could you take a pass please?

**Comment by simonlynen** [comment]
@idryanov ptal

File: nabo/nabo.h [hunk 1]
```diff
@@ -266,6 +268,7 @@ namespace Nabo
 		//! the high bound of the search space (axis-aligned bounding box)
 		const Vector maxBound;
 		
+
```
Comment (idryanov) [inline nabo/nabo.h]: revert?

**Comment by idryanov** [comment]
LGTM. Do you have unit tests that check the behavior with the default type (MatrixXf)?

**Comment by simonlynen** [comment]
@HannesSommer @idryanov thanks for the review.

@idryanov For the 3d-data I test the types: ```MatrixXf```, ```Matrix3Xf``` and ```Map<const Matrix3Xf>``` [https://github.com/ethz-asl/libnabo/blob/feature/zero_copy/tests/knnvalidate.cpp#L243-L245](diff)

**Pull request merged**